
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://xin-huang.github.io/selscape/userguide/output/">
      
      
        <link rel="prev" href="../configuration/">
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.39">
    
    
      
        <title>Output - selscape</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.8c3ca2c6.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#output" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="selscape" class="md-header__button md-logo" aria-label="selscape" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            selscape
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Output
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/xin-huang/selscape" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="selscape" class="md-nav__button md-logo" aria-label="selscape" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    selscape
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/xin-huang/selscape" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    User Guide
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            User Guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Output
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Output
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#understanding-file-paths" class="md-nav__link">
    <span class="md-ellipsis">
      Understanding File Paths
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#positive-selection" class="md-nav__link">
    <span class="md-ellipsis">
      Positive Selection
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Positive Selection">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#selscan" class="md-nav__link">
    <span class="md-ellipsis">
      selscan
    </span>
  </a>
  
    <nav class="md-nav" aria-label="selscan">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#within-population-statistics-ihs-nsl" class="md-nav__link">
    <span class="md-ellipsis">
      Within-Population Statistics (iHS, nSL)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cross-population-statistics-xp-ehh-xp-nsl" class="md-nav__link">
    <span class="md-ellipsis">
      Cross-Population Statistics (XP-EHH, XP-nSL)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scikit-allel" class="md-nav__link">
    <span class="md-ellipsis">
      scikit-allel
    </span>
  </a>
  
    <nav class="md-nav" aria-label="scikit-allel">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#windowed-tajimas-d" class="md-nav__link">
    <span class="md-ellipsis">
      Windowed Tajima's D
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#moving-tajimas-d" class="md-nav__link">
    <span class="md-ellipsis">
      Moving Tajima's D
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#delta-tajimas-d-cross-population" class="md-nav__link">
    <span class="md-ellipsis">
      Delta Tajima's D (Cross-Population)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#balancing-selection" class="md-nav__link">
    <span class="md-ellipsis">
      Balancing Selection
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Balancing Selection">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#betascan" class="md-nav__link">
    <span class="md-ellipsis">
      BetaScan
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scikit-allel_1" class="md-nav__link">
    <span class="md-ellipsis">
      scikit-allel
    </span>
  </a>
  
    <nav class="md-nav" aria-label="scikit-allel">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#windowed-tajimas-d_1" class="md-nav__link">
    <span class="md-ellipsis">
      Windowed Tajima's D
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#moving-tajimas-d_1" class="md-nav__link">
    <span class="md-ellipsis">
      Moving Tajima's D
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#distribution-of-fitness-effects" class="md-nav__link">
    <span class="md-ellipsis">
      Distribution of Fitness Effects
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Distribution of Fitness Effects">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dadidadi-cli" class="md-nav__link">
    <span class="md-ellipsis">
      dadi/dadi-cli
    </span>
  </a>
  
    <nav class="md-nav" aria-label="dadi/dadi-cli">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#demographic-model-inference" class="md-nav__link">
    <span class="md-ellipsis">
      Demographic Model Inference
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#demographic-model-fit" class="md-nav__link">
    <span class="md-ellipsis">
      Demographic Model Fit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#distribution-of-fitness-effects_1" class="md-nav__link">
    <span class="md-ellipsis">
      Distribution of Fitness Effects
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dfe-model-fit-plot" class="md-nav__link">
    <span class="md-ellipsis">
      DFE Model Fit Plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mutation-proportion-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Mutation Proportion Plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#confidence-intervals" class="md-nav__link">
    <span class="md-ellipsis">
      Confidence Intervals
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#functional-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Functional Annotation
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Functional Annotation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annovar-annotated-outliers" class="md-nav__link">
    <span class="md-ellipsis">
      ANNOVAR Annotated Outliers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#outlier-gene-lists" class="md-nav__link">
    <span class="md-ellipsis">
      Outlier Gene Lists
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#go-enrichment-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      GO Enrichment Analysis
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GO Enrichment Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gowinda-enrichment-results" class="md-nav__link">
    <span class="md-ellipsis">
      Gowinda Enrichment Results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#enrichment-visualization" class="md-nav__link">
    <span class="md-ellipsis">
      Enrichment Visualization
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#interactive-html-reports" class="md-nav__link">
    <span class="md-ellipsis">
      Interactive HTML Reports
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Overview">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#understanding-file-paths" class="md-nav__link">
    <span class="md-ellipsis">
      Understanding File Paths
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#positive-selection" class="md-nav__link">
    <span class="md-ellipsis">
      Positive Selection
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Positive Selection">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#selscan" class="md-nav__link">
    <span class="md-ellipsis">
      selscan
    </span>
  </a>
  
    <nav class="md-nav" aria-label="selscan">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#within-population-statistics-ihs-nsl" class="md-nav__link">
    <span class="md-ellipsis">
      Within-Population Statistics (iHS, nSL)
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#cross-population-statistics-xp-ehh-xp-nsl" class="md-nav__link">
    <span class="md-ellipsis">
      Cross-Population Statistics (XP-EHH, XP-nSL)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scikit-allel" class="md-nav__link">
    <span class="md-ellipsis">
      scikit-allel
    </span>
  </a>
  
    <nav class="md-nav" aria-label="scikit-allel">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#windowed-tajimas-d" class="md-nav__link">
    <span class="md-ellipsis">
      Windowed Tajima's D
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#moving-tajimas-d" class="md-nav__link">
    <span class="md-ellipsis">
      Moving Tajima's D
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#delta-tajimas-d-cross-population" class="md-nav__link">
    <span class="md-ellipsis">
      Delta Tajima's D (Cross-Population)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#balancing-selection" class="md-nav__link">
    <span class="md-ellipsis">
      Balancing Selection
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Balancing Selection">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#betascan" class="md-nav__link">
    <span class="md-ellipsis">
      BetaScan
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#scikit-allel_1" class="md-nav__link">
    <span class="md-ellipsis">
      scikit-allel
    </span>
  </a>
  
    <nav class="md-nav" aria-label="scikit-allel">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#windowed-tajimas-d_1" class="md-nav__link">
    <span class="md-ellipsis">
      Windowed Tajima's D
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#moving-tajimas-d_1" class="md-nav__link">
    <span class="md-ellipsis">
      Moving Tajima's D
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#distribution-of-fitness-effects" class="md-nav__link">
    <span class="md-ellipsis">
      Distribution of Fitness Effects
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Distribution of Fitness Effects">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#dadidadi-cli" class="md-nav__link">
    <span class="md-ellipsis">
      dadi/dadi-cli
    </span>
  </a>
  
    <nav class="md-nav" aria-label="dadi/dadi-cli">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#demographic-model-inference" class="md-nav__link">
    <span class="md-ellipsis">
      Demographic Model Inference
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#demographic-model-fit" class="md-nav__link">
    <span class="md-ellipsis">
      Demographic Model Fit
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#distribution-of-fitness-effects_1" class="md-nav__link">
    <span class="md-ellipsis">
      Distribution of Fitness Effects
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#dfe-model-fit-plot" class="md-nav__link">
    <span class="md-ellipsis">
      DFE Model Fit Plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#mutation-proportion-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Mutation Proportion Plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#confidence-intervals" class="md-nav__link">
    <span class="md-ellipsis">
      Confidence Intervals
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#functional-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Functional Annotation
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Functional Annotation">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#annovar-annotated-outliers" class="md-nav__link">
    <span class="md-ellipsis">
      ANNOVAR Annotated Outliers
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#outlier-gene-lists" class="md-nav__link">
    <span class="md-ellipsis">
      Outlier Gene Lists
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#go-enrichment-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      GO Enrichment Analysis
    </span>
  </a>
  
    <nav class="md-nav" aria-label="GO Enrichment Analysis">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gowinda-enrichment-results" class="md-nav__link">
    <span class="md-ellipsis">
      Gowinda Enrichment Results
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#enrichment-visualization" class="md-nav__link">
    <span class="md-ellipsis">
      Enrichment Visualization
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#interactive-html-reports" class="md-nav__link">
    <span class="md-ellipsis">
      Interactive HTML Reports
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="output">Output</h1>
<h2 id="overview">Overview</h2>
<p><code>selscape</code> organizes results by analysis type and population:</p>
<div class="codehilite"><pre><span></span><code>results/
├── positive_selection/
│   ├── selscan/
│   │   ├── {species}/1pop/{ppl}/{method}_{maf}/     # Within-population (iHS, nSL)
│   │   └── {species}/2pop/{pair}/{method}_{maf}/    # Cross-population (XP-EHH, XP-nSL)
│   └── scikit-allel/
│       ├── {species}/1pop/{ppl}/{method}/{window}_{step}/   # Negative Tajima&#39;s D
│       └── {species}/2pop/{pair}/{method}/{window}_{step}/  # Delta Tajima&#39;s D
├── balancing_selection/
│   ├── betascan/{species}/{ppl}/m_{core_frq}/
│   └── scikit-allel/{species}/{method}/{ppl}/{window}_{step}/  # Positive Tajima&#39;s D
├── dadi/{species}/dfe/{ppl}/
│   ├── InferDM/        # Demographic models
│   ├── InferDFE/       # DFE inference
│   ├── StatDFE/        # Confidence intervals
│   └── plots/          # Fitted model visualizations
</code></pre></div>

<h3 id="understanding-file-paths">Understanding File Paths</h3>
<p>Throughout this documentation, file paths use the following wildcards:</p>
<ul>
<li><code>{species}</code>: Species name (e.g., "Human")</li>
<li><code>{ppl}</code>: Population code (e.g., "YRI")</li>
<li><code>{pair}</code>: Population pair (e.g., "YRI_CHS")</li>
<li><code>{method}</code>: Selection statistic (e.g., "ihs", "windowed_tajima_d")</li>
<li><code>{maf}</code>: Minor allele frequency (e.g., "0.05")</li>
<li><code>{cutoff}</code>: Top proportion threshold (e.g., "0.0005")</li>
<li><code>{window}_{step}</code>: Window and step sizes</li>
<li><code>{ref_genome}</code>: Reference genome build (e.g., "hg38")</li>
<li><code>{demog}</code>: Demographic model (e.g., "two_epoch")</li>
<li><code>{dfe}</code>: DFE model (e.g., "lognormal")</li>
</ul>
<h2 id="positive-selection">Positive Selection</h2>
<h3 id="selscan">selscan</h3>
<h4 id="within-population-statistics-ihs-nsl">Within-Population Statistics (iHS, nSL)</h4>
<p><strong>File path:</strong></p>
<div class="codehilite"><pre><span></span><code>results/positive_selection/selscan/{species}/1pop/{ppl}/{method}_{maf}/{ppl}.normalized.{method}.maf_{maf}.top_{cutoff}.outliers.scores
</code></pre></div>

<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code>SNP             CHR     BP      normalized_nsl  abs_normalized_nsl
21:35702539     21      35702539        -4.88331        4.88331
20:57138931     20      57138931        4.68161         4.68161
21:28603177     21      28603177        4.63644         4.63644
</code></pre></div>

<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP</td>
<td>SNP identifier in format chromosome:position</td>
</tr>
<tr>
<td>CHR</td>
<td>Chromosome number</td>
</tr>
<tr>
<td>BP</td>
<td>Base pair position</td>
</tr>
<tr>
<td>normalized_{method}</td>
<td>Normalized selection statistic (iHS or nSL)</td>
</tr>
<tr>
<td>abs_normalized_{method}</td>
<td>Absolute value of normalized statistic</td>
</tr>
</tbody>
</table>
<p>Positive values may indicate selection favoring the derived allele, while negative values may indicate selection favoring the ancestral allele. High absolute values may indicate recent positive selection.</p>
<p><img alt="selscan Manhattan Plot" src="../../images/selscan_manhattan_plot.png" /></p>
<h4 id="cross-population-statistics-xp-ehh-xp-nsl">Cross-Population Statistics (XP-EHH, XP-nSL)</h4>
<p><strong>File path:</strong></p>
<div class="codehilite"><pre><span></span><code>results/positive_selection/selscan/{species}/2pop/{pair}/{method}_{maf}/{pair}.normalized.{method}.maf_{maf}.top_{cutoff}.outliers.scores
</code></pre></div>

<p>Same format as within-population statistics.</p>
<p>Positive values may indicate selection in the first population (e.g., YRI in YRI_CHS), while negative values may indicate selection in the second population (e.g., CHS in YRI_CHS).</p>
<p>More information: <a href="https://github.com/szpiech/selscan">selscan</a>.</p>
<h3 id="scikit-allel">scikit-allel</h3>
<h4 id="windowed-tajimas-d">Windowed Tajima's D</h4>
<p><strong>File path:</strong></p>
<div class="codehilite"><pre><span></span><code>results/positive_selection/scikit-allel/{species}/1pop/{ppl}/windowed_tajima/{window}_{step}/{ppl}.windowed_tajima..top_{cutoff}.outliers.scores
</code></pre></div>

<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code>SNP             CHR     BP      tajima_d        window_start    window_end      n_snps
20:61849        20      61849   -0.5400134      61849           161848          215
20:261849       20      261849  -0.2830888      261849          361848          400
</code></pre></div>

<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP</td>
<td>SNP identifier in format chromosome:position</td>
</tr>
<tr>
<td>CHR</td>
<td>Chromosome number</td>
</tr>
<tr>
<td>BP</td>
<td>Base pair position (window center)</td>
</tr>
<tr>
<td>tajima_d</td>
<td>Tajima's D statistic</td>
</tr>
<tr>
<td>window_start</td>
<td>Window start position in base pairs</td>
</tr>
<tr>
<td>window_end</td>
<td>Window end position in base pairs</td>
</tr>
<tr>
<td>n_snps</td>
<td>Number of SNPs in window</td>
</tr>
</tbody>
</table>
<p>Negative values suggest recent positive selection (selective sweeps), where beneficial alleles rapidly increased in frequency.</p>
<p><img alt="Windowed Tajima's D Manhattan Plot" src="../../images/windowed_tajima_d_manhattan_plot.png" /></p>
<p>For all Tajima's D methods, each data point in the Manhattan plot represents the starting position (<code>window_start</code>) of a window.</p>
<h4 id="moving-tajimas-d">Moving Tajima's D</h4>
<p><strong>File path:</strong></p>
<div class="codehilite"><pre><span></span><code>results/positive_selection/scikit-allel/{species}/1pop/{ppl}/moving_tajima_d/{window}_{step}/{ppl}.moving_tajima_d.top_{cutoff}.outliers.scores
</code></pre></div>

<p>Similar format to windowed approach, but windows are defined by SNP counts rather than base pairs.</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP</td>
<td>SNP identifier in format chromosome:position</td>
</tr>
<tr>
<td>CHR</td>
<td>Chromosome number</td>
</tr>
<tr>
<td>BP</td>
<td>Base pair position (window center)</td>
</tr>
<tr>
<td>tajima_d</td>
<td>Tajima's D statistic</td>
</tr>
<tr>
<td>window_start</td>
<td>Window start position in base pairs</td>
</tr>
<tr>
<td>window_end</td>
<td>Window end position in base pairs</td>
</tr>
<tr>
<td>n_snps</td>
<td>Number of SNPs in window (constant)</td>
</tr>
</tbody>
</table>
<p><img alt="Moving Tajima's D Manhattan Plot" src="../../images/moving_tajima_d_manhattan_plot.png" /></p>
<h4 id="delta-tajimas-d-cross-population">Delta Tajima's D (Cross-Population)</h4>
<p><strong>File path:</strong></p>
<div class="codehilite"><pre><span></span><code>results/positive_selection/scikit-allel/{species}/2pop/{pair}/delta_moving_tajima_d/{window}_{step}/{pair}.delta_moving_tajima_d.top_{cutoff}.outliers.scores
</code></pre></div>

<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code>SNP             CHR     BP      delta_tajima_d  window_start    window_end      n_snps  abs_delta_tajima_d
21:26659162     21      26659162        3.41144         26659162        26674022        100     3.41144
21:21456001     21      21456001        3.17360         21456001        21473170        100     3.17360
21:27242333     21      27242333        -3.07313        27242333        27258823        100     3.07313
</code></pre></div>

<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP</td>
<td>SNP identifier in format chromosome:position</td>
</tr>
<tr>
<td>CHR</td>
<td>Chromosome number</td>
</tr>
<tr>
<td>BP</td>
<td>Base pair position (window center)</td>
</tr>
<tr>
<td>delta_tajima_d</td>
<td>Difference in Tajima's D between populations</td>
</tr>
<tr>
<td>window_start</td>
<td>Window start position in base pairs</td>
</tr>
<tr>
<td>window_end</td>
<td>Window end position in base pairs</td>
</tr>
<tr>
<td>n_snps</td>
<td>Number of SNPs in window</td>
</tr>
<tr>
<td>abs_delta_tajima_d</td>
<td>Absolute value of delta Tajima's D</td>
</tr>
</tbody>
</table>
<p><img alt="Delta Tajima's D Manhattan Plot" src="../../images/delta_tajima_manhattan_plot.png" /></p>
<p>Large positive or negative values indicate selection acting differently between populations.</p>
<p>More information: <a href="https://scikit-allel.readthedocs.io/en/stable/index.html">scikit-allel</a>.</p>
<h2 id="balancing-selection">Balancing Selection</h2>
<h3 id="betascan">BetaScan</h3>
<p><strong>File path:</strong></p>
<div class="codehilite"><pre><span></span><code>results/balancing_selection/betascan/{species}/{ppl}/m_{core_frq}/{ppl}.{ref_genome}.m_{core_frq}.b1.top_{cutoff}.outliers.scores
</code></pre></div>

<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code>SNP             CHR     BP      B1
20:1914695      20      1914695 41.939494
20:1914712      20      1914712 41.479453
20:1914729      20      1914729 40.555049
</code></pre></div>

<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP</td>
<td>SNP identifier in format chromosome:position</td>
</tr>
<tr>
<td>CHR</td>
<td>Chromosome number</td>
</tr>
<tr>
<td>BP</td>
<td>Base pair position</td>
</tr>
<tr>
<td>B1</td>
<td>Beta1 statistic for balancing selection</td>
</tr>
</tbody>
</table>
<p>Higher B1 values indicate stronger evidence for long-term balancing selection, where multiple alleles are maintained over extended periods.</p>
<p><img alt="BetaScan Manhattan Plot" src="../../images/betascan_manhattan_plot.png" /></p>
<p>More information: <a href="https://github.com/ksiewert/BetaScan/wiki">BetaScan</a>.</p>
<h3 id="scikit-allel_1">scikit-allel</h3>
<h4 id="windowed-tajimas-d_1">Windowed Tajima's D</h4>
<p><strong>File path:</strong></p>
<div class="codehilite"><pre><span></span><code>results/balancing_selection/scikit-allel/{species}/windowed_tajima_d/{ppl}/{window}_{step}/{ppl}.windowed_tajima_d.top_{cutoff}.outliers.scores
</code></pre></div>

<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code>SNP             CHR     BP      tajima_d        window_start    window_end      n_snps
20:61849        20      61849   -0.5400134      61849           161848          215
20:261849       20      261849  -0.2830888      261849          361848          400
</code></pre></div>

<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP</td>
<td>SNP identifier in format chromosome:position</td>
</tr>
<tr>
<td>CHR</td>
<td>Chromosome number</td>
</tr>
<tr>
<td>BP</td>
<td>Base pair position (window center)</td>
</tr>
<tr>
<td>tajima_d</td>
<td>Tajima's D statistic</td>
</tr>
<tr>
<td>window_start</td>
<td>Window start position in base pairs</td>
</tr>
<tr>
<td>window_end</td>
<td>Window end position in base pairs</td>
</tr>
<tr>
<td>n_snps</td>
<td>Number of SNPs in window</td>
</tr>
</tbody>
</table>
<p>Positive values indicate balancing selection, where multiple alleles are maintained in the population.</p>
<h4 id="moving-tajimas-d_1">Moving Tajima's D</h4>
<p><strong>File path:</strong>  </p>
<div class="codehilite"><pre><span></span><code>results/balancing_selection/scikit-allel/{species}/moving_tajima_d/{window}_{step}/{ppl}.moving_tajima_d.top_{cutoff}.outliers.scores
</code></pre></div>

<p>Similar format to windowed approach, but windows are defined by SNP counts rather than base pairs.</p>
<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP</td>
<td>SNP identifier in format chromosome:position</td>
</tr>
<tr>
<td>CHR</td>
<td>Chromosome number</td>
</tr>
<tr>
<td>BP</td>
<td>Base pair position (window center)</td>
</tr>
<tr>
<td>tajima_d</td>
<td>Tajima's D statistic</td>
</tr>
<tr>
<td>window_start</td>
<td>Window start position in base pairs</td>
</tr>
<tr>
<td>window_end</td>
<td>Window end position in base pairs</td>
</tr>
<tr>
<td>n_snps</td>
<td>Number of SNPs in window (constant)</td>
</tr>
</tbody>
</table>
<h2 id="distribution-of-fitness-effects">Distribution of Fitness Effects</h2>
<h3 id="dadidadi-cli">dadi/dadi-cli</h3>
<h4 id="demographic-model-inference">Demographic Model Inference</h4>
<p><strong>File path:</strong> <code>results/dadi/{species}/dfe/{ppl}/InferDM/{ppl}.{ref_genome}.{demog}.InferDM.bestfits</code></p>
<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code># Log(likelihood)       nu      T       misid   theta
-29.738235902874877     3.2402258068957983      0.13580062396208237     0.0010470465761291449   321.237660524018
-29.73823629293304      3.2417998728753408      0.13578140459967458     0.0010576355371547388   321.23238379628344
-29.73823658527745      3.237659010918116       0.13591524984020853     0.0010605026099018222   321.22253191565875
</code></pre></div>

<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Log(likelihood)</td>
<td>Log-likelihood of the model fit (higher is better)</td>
</tr>
<tr>
<td>nu</td>
<td>Ratio of current population size to ancestral population size</td>
</tr>
<tr>
<td>T</td>
<td>Time since population size change, measured in 2Ne generations</td>
</tr>
<tr>
<td>misid</td>
<td>Proportion of misidentified ancestral states</td>
</tr>
<tr>
<td>theta</td>
<td>Population-scaled mutation rate (4Neμ)</td>
</tr>
</tbody>
</table>
<h4 id="demographic-model-fit">Demographic Model Fit</h4>
<p><strong>File path:</strong> <code>results/dadi/{species}/dfe/{ppl}/plots/{ppl}.{ref_genome}.{demog}.fitted.png</code></p>
<p><img alt="Demographic Model Fit" src="../../images/two_epoch.png" /></p>
<p>The visualization shows the observed frequency spectrum (data) compared to the model-predicted spectrum.</p>
<h4 id="distribution-of-fitness-effects_1">Distribution of Fitness Effects</h4>
<p><strong>File path:</strong> <code>results/dadi/{species}/dfe/{ppl}/InferDFE/{ppl}.{ref_genome}.{demog}.{dfe}.InferDFE.bestfits</code></p>
<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code><span class="gh">#</span> Log(likelihood)       log_mu  log_sigma       misid   theta
-28.737041385073574     2.5282728150148785      4.591179583286655       0.039371652920637594    742.0589958104816
-28.73804138508753      2.5282735377291865      4.591183662610013       0.0393715391668421      742.0589958104816
-28.737041388380618     2.5283038170363823      4.5912804442420105      0.039370931031321235    742.0589958104816
</code></pre></div>

<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Log(likelihood)</td>
<td>Log-likelihood of the DFE model fit (higher is better)</td>
</tr>
<tr>
<td>log_mu</td>
<td>Mean of log-normal distribution of selection coefficients</td>
</tr>
<tr>
<td>log_sigma</td>
<td>Standard deviation of log-normal distribution of selection coefficients</td>
</tr>
<tr>
<td>misid</td>
<td>Proportion of misidentified ancestral states</td>
</tr>
<tr>
<td>theta</td>
<td>Population-scaled mutation rate (4Neμ)</td>
</tr>
</tbody>
</table>
<h4 id="dfe-model-fit-plot">DFE Model Fit Plot</h4>
<p><strong>File path:</strong> <code>results/dadi/{species}/dfe/{ppl}/plots/{ppl}.{ref_genome}.{demog}.{dfe}.fitted.png</code></p>
<p><img alt="DFE Model Fit" src="../../images/two_epoch.lognormal.png" />
The visualization shows the observed vs. model-predicted frequency spectrum for nonsynonymous mutations.</p>
<h4 id="mutation-proportion-plot">Mutation Proportion Plot</h4>
<p><strong>File path:</strong> <code>results/dadi/{species}/dfe/{ppl}/plots/{ppl}.{ref_genome}.{demog}.{dfe}.fitted.mut.prop.png</code></p>
<p><img alt="Mutation Proportions" src="../../images/two_epoch.lognormal.mut.prop.png" /></p>
<p>The mutation proportion plot shows the distribution of mutations by selection strength (2N|s|): nearly neutral mutations (&lt; 1), weakly deleterious mutations (1–10), moderately deleterious mutations (10–100), and strongly deleterious mutations (&gt; 100).</p>
<h4 id="confidence-intervals">Confidence Intervals</h4>
<p><strong>File path:</strong>  <code>results/dadi/{species}/dfe/{ppl}/StatDFE/{ppl}.{ref_genome}.{demog}.{dfe}.godambe.ci</code></p>
<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code>Estimated 95% uncerts (theta adj), with step size 0.1): [2.45625971 3.17513531 0.11405905]
Lower bounds of 95% confidence interval : [-0.65471712 -0.74208004 -0.0046111 ]
Upper bounds of 95% confidence interval : [4.2578023  5.60819059 0.22350701]

Estimated 95% uncerts (theta adj), with step size 0.01): [2.19800876 2.55236006 0.10687941]
Lower bounds of 95% confidence interval : [-0.39646617 -0.11930478  0.00256854]
Upper bounds of 95% confidence interval : [3.99955135 4.98541534 0.21632737]

Estimated 95% uncerts (theta adj), with step size 0.001): [2.19575422 2.54701563 0.10682199]
Lower bounds of 95% confidence interval : [-0.39421163 -0.11396035  0.00262596]
Upper bounds of 95% confidence interval : [3.99729681 4.98007091 0.21626994]
</code></pre></div>

<p>Godambe-based 95% confidence intervals for all DFE parameters.</p>
<p>More information: <a href="https://dadi.readthedocs.io/en/latest/">dadi</a> and <a href="https://dadi-cli.readthedocs.io/en/latest/">dadi-cli</a>.</p>
<h2 id="functional-annotation">Functional Annotation</h2>
<p>Gene and functional annotation of outliers:</p>
<h3 id="annovar-annotated-outliers">ANNOVAR Annotated Outliers</h3>
<p>All outliers are annotated with gene information:</p>
<p><strong>File path:</strong> <code>{analysis_path}/{file_prefix}.annotated.outliers</code></p>
<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code><span class="nv">Chr</span><span class="w">     </span><span class="nv">Start</span><span class="w">   </span><span class="k">End</span><span class="w">     </span><span class="nv">Ref</span><span class="w">     </span><span class="nv">Alt</span><span class="w">     </span><span class="nv">Func</span>.<span class="nv">refGene</span><span class="w">    </span><span class="nv">Gene</span>.<span class="nv">refGene</span><span class="w">    </span><span class="nv">ExonicFunc</span>.<span class="nv">refGene</span><span class="w">      </span><span class="nv">AAChange</span>.<span class="nv">refGene</span>
<span class="mi">20</span><span class="w">      </span><span class="mi">14150039</span><span class="w">        </span><span class="mi">14150039</span><span class="w">        </span><span class="nv">C</span><span class="w">       </span><span class="nv">A</span><span class="w">       </span><span class="nv">intronic</span><span class="w">        </span><span class="nv">MACROD2</span><span class="w"> </span>.<span class="w">       </span>.
<span class="mi">20</span><span class="w">      </span><span class="mi">17148100</span><span class="w">        </span><span class="mi">17148100</span><span class="w">        </span><span class="nv">C</span><span class="w">       </span><span class="nv">A</span><span class="w">       </span><span class="nv">intergenic</span><span class="w">      </span><span class="nv">OTOR</span><span class="c1">;PCSK2      .       .</span>
<span class="mi">20</span><span class="w">      </span><span class="mi">23595497</span><span class="w">        </span><span class="mi">23595497</span><span class="w">        </span><span class="nv">G</span><span class="w">       </span><span class="nv">A</span><span class="w">       </span><span class="nv">intergenic</span><span class="w">      </span><span class="nv">CST9L</span><span class="c1">;CST9      .       .</span>
</code></pre></div>

<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Chr</td>
<td>Chromosome number</td>
</tr>
<tr>
<td>Start</td>
<td>Variant start position in base pairs</td>
</tr>
<tr>
<td>End</td>
<td>Variant end position in base pairs</td>
</tr>
<tr>
<td>Ref</td>
<td>Reference allele</td>
</tr>
<tr>
<td>Alt</td>
<td>Alternative allele</td>
</tr>
<tr>
<td>Func.refGene</td>
<td>Functional region (exonic, intronic, intergenic, UTR, etc.)</td>
</tr>
<tr>
<td>Gene.refGene</td>
<td>Gene symbol(s)</td>
</tr>
<tr>
<td>ExonicFunc.refGene</td>
<td>Exonic function (synonymous, nonsynonymous, stopgain, stoploss, etc.)</td>
</tr>
<tr>
<td>AAChange.refGene</td>
<td>Amino acid change for coding variants</td>
</tr>
</tbody>
</table>
<h3 id="outlier-gene-lists">Outlier Gene Lists</h3>
<p><strong>File path:</strong> <code>{analysis_path}/{file_prefix}.outlier.genes</code></p>
<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code>Gene
HLCS
MACROD2
PCSK2
UQCC1
</code></pre></div>

<p>Simple list of unique genes extracted from annotated outliers.</p>
<p>More information: <a href="https://annovar.openbioinformatics.org/en/latest/">ANNOVAR</a>.</p>
<h2 id="go-enrichment-analysis">GO Enrichment Analysis</h2>
<p>Statistical testing for GO enrichment in outlier genes:</p>
<h3 id="gowinda-enrichment-results">Gowinda Enrichment Results</h3>
<p><strong>File path:</strong>  <code>{analysis_path}/{file_prefix}.gowinda.enrichment.tsv</code></p>
<p><strong>Example output:</strong></p>
<div class="codehilite"><pre><span></span><code><span class="n">GO_ID</span><span class="w">       </span><span class="n">avg_genes_sim</span><span class="w">   </span><span class="n">genes_found</span><span class="w">     </span><span class="n">p_value</span><span class="w">         </span><span class="n">p_adjusted</span><span class="w">      </span><span class="n">genes_uniq</span><span class="w">      </span><span class="n">genes_max</span><span class="w">       </span><span class="n">genes_total</span><span class="w">     </span><span class="n">description</span><span class="w">                     </span><span class="n">gene_list</span>
<span class="k">GO</span><span class="err">:</span><span class="mi">0034551</span><span class="w">  </span><span class="mf">0.008</span><span class="w">           </span><span class="mi">1</span><span class="w">               </span><span class="mf">0.0082240000</span><span class="w">    </span><span class="mf">1.0000000000</span><span class="w">    </span><span class="mi">1</span><span class="w">               </span><span class="mi">1</span><span class="w">               </span><span class="mi">14</span><span class="w">              </span><span class="n">mitochondrial</span><span class="w"> </span><span class="n">complex</span><span class="w"> </span><span class="n">III</span><span class="w">       </span><span class="n">uqcc1</span>
<span class="k">GO</span><span class="err">:</span><span class="mi">0031410</span><span class="w">  </span><span class="mf">0.270</span><span class="w">           </span><span class="mi">2</span><span class="w">               </span><span class="mf">0.0247750000</span><span class="w">    </span><span class="mf">1.0000000000</span><span class="w">    </span><span class="mi">2</span><span class="w">               </span><span class="mi">37</span><span class="w">              </span><span class="mi">1017</span><span class="w">            </span><span class="n">cytoplasmic</span><span class="w"> </span><span class="n">vesicle</span><span class="w">             </span><span class="n">uqcc1</span><span class="p">,</span><span class="n">pcsk2</span>
<span class="k">GO</span><span class="err">:</span><span class="mi">0004077</span><span class="w">  </span><span class="mf">0.030</span><span class="w">           </span><span class="mi">1</span><span class="w">               </span><span class="mf">0.0297130000</span><span class="w">    </span><span class="mf">1.0000000000</span><span class="w">    </span><span class="mi">1</span><span class="w">               </span><span class="mi">1</span><span class="w">               </span><span class="mi">1</span><span class="w">               </span><span class="n">biotin</span><span class="w"> </span><span class="n">ligase</span><span class="w"> </span><span class="n">activity</span><span class="w">          </span><span class="n">hlcs</span>
</code></pre></div>

<table>
<thead>
<tr>
<th>Column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>GO_ID</td>
<td>Gene Ontology term identifier</td>
</tr>
<tr>
<td>avg_genes_sim</td>
<td>Average number of genes in simulations</td>
</tr>
<tr>
<td>genes_found</td>
<td>Number of outlier genes found in this GO term</td>
</tr>
<tr>
<td>p_value</td>
<td>Uncorrected p-value from simulation</td>
</tr>
<tr>
<td>p_adjusted</td>
<td>FDR-corrected p-value</td>
</tr>
<tr>
<td>genes_uniq</td>
<td>Number of unique genes in this GO term</td>
</tr>
<tr>
<td>genes_max</td>
<td>Maximum possible genes for this term</td>
</tr>
<tr>
<td>genes_total</td>
<td>Total genes in background set</td>
</tr>
<tr>
<td>description</td>
<td>GO term description</td>
</tr>
<tr>
<td>gene_list</td>
<td>Comma-separated list of outlier genes</td>
</tr>
</tbody>
</table>
<h3 id="enrichment-visualization">Enrichment Visualization</h3>
<p><strong>File path:</strong>  <code>{analysis_path}/{file_prefix}.gowinda.enrichment.png</code></p>
<p><img alt="Gowinda Enrichment Plot" src="../../images/gowinda_enrichment_plot.png" /></p>
<p>The enrichment plot shows top 20 GO terms, sorted by their significance (FDR adjusted p-value) and then by the number of genes found. Bars are colored based on significance: red for highly significant enrichment (&lt; 0.001), purple for significant enrichment (&lt; 0.01), blue for moderately significant enrichment (&lt; 0.05), and gray for non-significant results (≥ 0.05).</p>
<p>More information: <a href="https://sourceforge.net/p/gowinda/wiki/Tutorial">Gowinda Tutorial</a>.</p>
<h2 id="interactive-html-reports">Interactive HTML Reports</h2>
<p>All methods generate interactive HTML tables with search, sort, and export functionality:</p>
<ul>
<li><strong>Gene lists</strong>: <code>{analysis_path}/{file_prefix}.outlier.genes.html</code></li>
<li><strong>Enrichment results</strong>: <code>{analysis_path}/{file_prefix}.gowinda.enrichment.html</code></li>
<li>
<p><strong>Best-fit parameters</strong>:</p>
</li>
<li>
<p><code>results/dadi/{species}/dfe/{ppl}/html/{ppl}.{ref_genome}.{demog}.InferDM.top10.bestfits.html</code></p>
</li>
<li><code>results/dadi/{species}/dfe/{ppl}/html/{ppl}.{ref_genome}.{demog}.{dfe}.InferDFE.top10.bestfits.html</code></li>
<li><code>results/dadi/{species}/dfe/{ppl}/html/{ppl}.{ref_genome}.{demog}.{dfe}.godambe.ci.html</code></li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.525ec568.min.js"></script>
      
    
  </body>
</html>