
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://xin-huang.github.io/selscape/userguide/output/">
      
      
        <link rel="prev" href="../configuration/">
      
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.40">
    
    
      
        <title>Output - Selscape</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.8c3ca2c6.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#output" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="Selscape" class="md-header__button md-logo" aria-label="Selscape" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Selscape
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Output
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/xin-huang/selscape" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="Selscape" class="md-nav__button md-logo" aria-label="Selscape" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    Selscape
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/xin-huang/selscape" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    User Guide
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            User Guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Output
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Output
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#betascan-results-balancing-selection" class="md-nav__link">
    <span class="md-ellipsis">
      Betascan Results (Balancing Selection)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#selscan-results-positive-selection" class="md-nav__link">
    <span class="md-ellipsis">
      Selscan Results (Positive Selection)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#selection-candidates-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Selection Candidates Annotation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dadi-cli-results-dfe" class="md-nav__link">
    <span class="md-ellipsis">
      Dadi-cli Results (DFE)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gowinda-enrichment-results" class="md-nav__link">
    <span class="md-ellipsis">
      Gowinda Enrichment Results
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#betascan-results-balancing-selection" class="md-nav__link">
    <span class="md-ellipsis">
      Betascan Results (Balancing Selection)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#selscan-results-positive-selection" class="md-nav__link">
    <span class="md-ellipsis">
      Selscan Results (Positive Selection)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#selection-candidates-annotation" class="md-nav__link">
    <span class="md-ellipsis">
      Selection Candidates Annotation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dadi-cli-results-dfe" class="md-nav__link">
    <span class="md-ellipsis">
      Dadi-cli Results (DFE)
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gowinda-enrichment-results" class="md-nav__link">
    <span class="md-ellipsis">
      Gowinda Enrichment Results
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="output">Output</h1>
<h2 id="betascan-results-balancing-selection">Betascan Results (Balancing Selection)</h2>
<p>Results are saved in <code>results/plots/betascan/Human/YRI/m_0.15/YRI.b1.top.0.0005.candidates.scores</code>.</p>
<div class="codehilite"><pre><span></span><code>SNP     CHR     BP      B1
22:22350476     22      22350476        29.950265
22:22350435     22      22350435        29.8842
22:22350418     22      22350418        29.332644
22:22350419     22      22350419        29.332644
22:22350497     22      22350497        29.327195
</code></pre></div>

<table>
<thead>
<tr>
<th>column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP</td>
<td>SNP identifier in format chromosome:position</td>
</tr>
<tr>
<td>CHR</td>
<td>Chromosome number</td>
</tr>
<tr>
<td>BP</td>
<td>Base pair position</td>
</tr>
<tr>
<td>B1</td>
<td>Beta1 statistic for balancing selection</td>
</tr>
</tbody>
</table>
<p>Higher B1 values suggests stronger evidence for balancing selection of loci where multiple alleles are maintained in the population over long periods.</p>
<p><img alt="Betascan Plot" src="../images/betascan_manhattan_plot.png" /></p>
<h2 id="selscan-results-positive-selection">Selscan Results (Positive Selection)</h2>
<p>Results are saved in <code>results/plots/selscan/Human/1pop/YRI/nsl_0.05/YRI.normalized.nsl.top.0.0005.candidates.scores</code>.</p>
<div class="codehilite"><pre><span></span><code>SNP     CHR     BP      normalized_nsl  abs_normalized_nsl
21:27635622     21      27635622        4.98329 4.98329
22:19879946     22      19879946        4.5454  4.5454
21:29707889     21      29707889        4.54516 4.54516
21:27617315     21      27617315        4.37302 4.37302
21:27642578     21      27642578        4.33443 4.33443
</code></pre></div>

<table>
<thead>
<tr>
<th>column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>SNP</td>
<td>SNP identifier in format chromosome:position</td>
</tr>
<tr>
<td>CHR</td>
<td>Chromosome number</td>
</tr>
<tr>
<td>BP</td>
<td>Base pair position</td>
</tr>
<tr>
<td>normalized_{method}</td>
<td>Normalized selection statistic (iHS, nSL, XP-EHH, XP-nSL)</td>
</tr>
</tbody>
</table>
<p>High scores indicate recent positive selection where beneficial alleles rapidly increased in frequency. With iHS, nSL statistics, positive values means selection favoring the derived allele, while negative values indicates that selection favors the  ancestral allele. With cross selection statistisc (XP-EHH, XP-nSL), positive values show selection in the population 1 (first population listed). Negative values suggest selection in population 2 (second population listed)</p>
<p><img alt="Selscan Plot" src="../images/selscan_manhattan_plot.png" /></p>
<h2 id="selection-candidates-annotation">Selection Candidates Annotation</h2>
<p><strong>Selscan Example</strong></p>
<p>Results are stored in <code>results/plots/selscan/Human/1pop/YRI/nsl_0.05/YRI.normalized.nsl.top.0.0005.annotated.candidates</code>.</p>
<div class="codehilite"><pre><span></span><code>Chr     Start   End     Ref     Alt     Func.refGene    Gene.refGene    GeneDetail.refGene      ExonicFunc.refGene      AAChange.refGene        normalized_nsl
21      19259240        19259240        A       C       intergenic      MIR548XHG;LINC01683     dist=499428;dist=634039 .       .       3.95068
21      19261015        19261015        A       G       intergenic      MIR548XHG;LINC01683     dist=501203;dist=632264 .       .       4.01447
21      19267995        19267995        T       C       intergenic      MIR548XHG;LINC01683     dist=508183;dist=625284 .       .       4.01979
21      19269299        19269299        T       G       intergenic      MIR548XHG;LINC01683     dist=509487;dist=623980 .       .       4.11579
21      26270854        26270854        G       T       intergenic      APP;CYYR1-AS1   dist=99726;dist=122781  .       .       4.05708
</code></pre></div>

<p>Betascan candidate results have the same format, except the last column is replaced with B1 score.</p>
<table>
<thead>
<tr>
<th>column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Chr</td>
<td>Chromosome</td>
</tr>
<tr>
<td>Start</td>
<td>Start Position of SNP window</td>
</tr>
<tr>
<td>End</td>
<td>End position of SNP window</td>
</tr>
<tr>
<td>Ref</td>
<td>Reference Allele</td>
</tr>
<tr>
<td>Alt</td>
<td>Alternative Allele</td>
</tr>
<tr>
<td>Func.refGene</td>
<td>Functional annotation (exonic, intronic, etc.)</td>
</tr>
<tr>
<td>Gene.refGene</td>
<td>Gene symbol</td>
</tr>
<tr>
<td>ExonicFunc.refGene</td>
<td>Exonic function (synonymous, nonsynonymous, etc.)</td>
</tr>
<tr>
<td>AAChange.refGene</td>
<td>Amino acid change</td>
</tr>
<tr>
<td>score</td>
<td>Selection statistic value</td>
</tr>
</tbody>
</table>
<h2 id="dadi-cli-results-dfe">Dadi-cli Results (DFE)</h2>
<p><strong>Distribution of Fitness</strong></p>
<p>Baboon Example (placeholder)</p>
<div class="codehilite"><pre><span></span><code><span class="gh">#</span> Log(likelihood)       log_mu  log_sigma       misid   theta
-1583.6419711050235     0.0001  10.036936601427616      0.03935425742951182     3146.354810257352
-1583.641978040827      0.0001  10.037430294509049      0.039352517953454194    3146.354810257352
-1583.6419830604987     0.0001  10.037590533056404      0.03933425868201494     3146.354810257352
-1583.6420065768962     0.0001  10.037898514671456      0.039345846769698795    3146.354810257352
-1583.642012206451      0.0001  10.036589176494077      0.039348107346255747    3146.354810257352
</code></pre></div>

<table>
<thead>
<tr>
<th>column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Log(likelihood)</td>
<td>mathematical measure how well model fits data</td>
</tr>
<tr>
<td>log_mu</td>
<td>Parameter describing the center of the lognormal distribution of selection coefficients</td>
</tr>
<tr>
<td>log_sigma</td>
<td>Parameter describing the spread of the lognormal distribution</td>
</tr>
<tr>
<td>misid</td>
<td>Proportion of misidentified ancestral states in the data</td>
</tr>
<tr>
<td>theta</td>
<td>Population-scaled mutation rate (4Nₑμ)</td>
</tr>
</tbody>
</table>
<p><strong>Demographic Model</strong></p>
<p>Baboon Example (placeholder)</p>
<div class="codehilite"><pre><span></span><code># Log(likelihood)       nu      T       misid   theta
-511.2556676138038      0.3072571458169096      0.029684712654180687    0.1867057098794246      15297.736007242134
-511.26099404545494     0.30821618471879975     0.029869554926965972    0.1866371110098524      15299.330420308932
-511.2617059355539      0.3046331410365202      0.02923435458096174     0.18693420973895863     15297.05292795244
-511.2556676138038      0.3072571458169096      0.029684712654180687    0.1867057098794246      15297.736007242134
-511.26099404545494     0.30821618471879975     0.029869554926965972    0.1866371110098524      15299.330420308932
</code></pre></div>

<table>
<thead>
<tr>
<th>column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Log(likelihood)</td>
<td>mathematical measure how well model fits data</td>
</tr>
<tr>
<td>nu</td>
<td>ratio of current population size to ancestral size</td>
</tr>
<tr>
<td>T (time)</td>
<td>time since population size changed, meassured in 2N generation</td>
</tr>
<tr>
<td>misid</td>
<td>error rate of data, accounting for misidentified ancestral alleles or seq error</td>
</tr>
<tr>
<td>theta</td>
<td>population scaled mutation rate</td>
</tr>
</tbody>
</table>
<h2 id="gowinda-enrichment-results">Gowinda Enrichment Results</h2>
<p>Results are saved in <code>results/functional_analysis_gowinda/Human/1pop/YRI/nsl_0.05/YRI.nsl_0.05.top.0.0005.gowinda.enrichment.tsv</code></p>
<div class="codehilite"><pre><span></span><code><span class="n">GO</span><span class="o">:</span><span class="mi">0006979</span><span class="w">  </span><span class="mf">0.072</span><span class="w">   </span><span class="mi">2</span><span class="w">   </span><span class="mf">0.0015570000</span><span class="w">    </span><span class="mf">0.1680687353</span><span class="w">    </span><span class="mi">2</span><span class="w">   </span><span class="mi">7</span><span class="w">   </span><span class="mi">161</span><span class="w"> </span><span class="n">involved_in</span><span class="w"> </span><span class="n">response</span><span class="w"> </span><span class="n">to</span><span class="w"> </span><span class="n">oxidative</span><span class="w"> </span><span class="n">stress</span><span class="w">    </span><span class="n">txnrd2</span><span class="o">,</span><span class="n">comt</span>
<span class="n">GO</span><span class="o">:</span><span class="mi">0042415</span><span class="w">  </span><span class="mf">0.007</span><span class="w">   </span><span class="mi">1</span><span class="w">   </span><span class="mf">0.0066270000</span><span class="w">    </span><span class="mf">0.1680687353</span><span class="w">    </span><span class="mi">1</span><span class="w">   </span><span class="mi">1</span><span class="w">   </span><span class="mi">10</span><span class="w">  </span><span class="n">involved_in</span><span class="w"> </span><span class="n">norepinephrine</span><span class="w"> </span><span class="n">metabolic</span><span class="w"> </span><span class="n">process</span><span class="w">    </span><span class="n">comt</span>
<span class="n">GO</span><span class="o">:</span><span class="mi">0042417</span><span class="w">  </span><span class="mf">0.007</span><span class="w">   </span><span class="mi">1</span><span class="w">   </span><span class="mf">0.0066270000</span><span class="w">    </span><span class="mf">0.1680687353</span><span class="w">    </span><span class="mi">1</span><span class="w">   </span><span class="mi">1</span><span class="w">   </span><span class="mi">18</span><span class="w">  </span><span class="n">involved_in</span><span class="w"> </span><span class="n">dopamine</span><span class="w"> </span><span class="n">metabolic</span><span class="w"> </span><span class="n">process</span><span class="w">  </span><span class="n">comt</span>
<span class="n">GO</span><span class="o">:</span><span class="mi">0002001</span><span class="w">  </span><span class="mf">0.007</span><span class="w">   </span><span class="mi">1</span><span class="w">   </span><span class="mf">0.0066270000</span><span class="w">    </span><span class="mf">0.1680687353</span><span class="w">    </span><span class="mi">1</span><span class="w">   </span><span class="mi">1</span><span class="w">   </span><span class="mi">4</span><span class="w">   </span><span class="n">acts_upstream_of_or_within</span><span class="w"> </span><span class="n">renin</span><span class="w"> </span><span class="n">secretion</span><span class="w"> </span><span class="n">into</span><span class="w"> </span><span class="n">blood</span><span class="w"> </span><span class="n">stream</span><span class="w">    </span><span class="n">comt</span>
<span class="n">GO</span><span class="o">:</span><span class="mi">0048243</span><span class="w">  </span><span class="mf">0.007</span><span class="w">   </span><span class="mi">1</span><span class="w">   </span><span class="mf">0.0066270000</span><span class="w">    </span><span class="mf">0.1680687353</span><span class="w">    </span><span class="mi">1</span><span class="w">   </span><span class="mi">1</span><span class="w">   </span><span class="mi">1</span><span class="w">   </span><span class="n">acts_upstream_of_or_within</span><span class="w"> </span><span class="n">norepinephrine</span><span class="w"> </span><span class="n">secretion</span><span class="w"> </span><span class="n">comt</span>
</code></pre></div>

<table>
<thead>
<tr>
<th>column</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>GO_ID</td>
<td>Gene Ontology term</td>
</tr>
<tr>
<td>avg_genes_sim</td>
<td>Average genes in simulations</td>
</tr>
<tr>
<td>genes_found</td>
<td>Genes found in candidates</td>
</tr>
<tr>
<td>p_value</td>
<td>Uncorrected p-value</td>
</tr>
<tr>
<td>p_adjusted</td>
<td>FDR-corrected p-value</td>
</tr>
<tr>
<td>genes_uniq</td>
<td>Unique genes in term</td>
</tr>
<tr>
<td>genes_max</td>
<td>Maximum genes possible</td>
</tr>
<tr>
<td>genes_total</td>
<td>Total genes in background</td>
</tr>
<tr>
<td>description</td>
<td>GO term description</td>
</tr>
<tr>
<td>gene_list</td>
<td>Comma-separated candidate genes</td>
</tr>
</tbody>
</table>
<p><img alt="Gowinda Enrichment plot (counts)" src="../images/gowinda_enrichment_plot.png" /></p>
<p>More information about gowinda can be found <a href="https://sourceforge.net/p/gowinda/wiki/Tutorial">here</a>.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.525ec568.min.js"></script>
      
    
  </body>
</html>