
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://xin-huang.github.io/selscape/userguide/configuration/">
      
      
        <link rel="prev" href="../usage/">
      
      
        <link rel="next" href="../output/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.39">
    
    
      
        <title>Configuration - selscape</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.8c3ca2c6.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#configuration" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="selscape" class="md-header__button md-logo" aria-label="selscape" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            selscape
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Configuration
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/xin-huang/selscape" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="selscape" class="md-nav__button md-logo" aria-label="selscape" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    selscape
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/xin-huang/selscape" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    User Guide
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            User Guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../usage/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#mainyaml" class="md-nav__link">
    <span class="md-ellipsis">
      main.yaml
    </span>
  </a>
  
    <nav class="md-nav" aria-label="main.yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#betascanyaml" class="md-nav__link">
    <span class="md-ellipsis">
      betascan.yaml
    </span>
  </a>
  
    <nav class="md-nav" aria-label="betascan.yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example_1" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_1" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#selscanyaml" class="md-nav__link">
    <span class="md-ellipsis">
      selscan.yaml
    </span>
  </a>
  
    <nav class="md-nav" aria-label="selscan.yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example_2" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_2" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scikit-allelyaml" class="md-nav__link">
    <span class="md-ellipsis">
      scikit-allel.yaml
    </span>
  </a>
  
    <nav class="md-nav" aria-label="scikit-allel.yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example_3" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_3" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dadi-cliyaml" class="md-nav__link">
    <span class="md-ellipsis">
      dadi-cli.yaml
    </span>
  </a>
  
    <nav class="md-nav" aria-label="dadi-cli.yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example_4" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_4" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#important-notes" class="md-nav__link">
    <span class="md-ellipsis">
      Important Notes
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../output/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Output
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#mainyaml" class="md-nav__link">
    <span class="md-ellipsis">
      main.yaml
    </span>
  </a>
  
    <nav class="md-nav" aria-label="main.yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#betascanyaml" class="md-nav__link">
    <span class="md-ellipsis">
      betascan.yaml
    </span>
  </a>
  
    <nav class="md-nav" aria-label="betascan.yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example_1" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_1" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#selscanyaml" class="md-nav__link">
    <span class="md-ellipsis">
      selscan.yaml
    </span>
  </a>
  
    <nav class="md-nav" aria-label="selscan.yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example_2" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_2" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#scikit-allelyaml" class="md-nav__link">
    <span class="md-ellipsis">
      scikit-allel.yaml
    </span>
  </a>
  
    <nav class="md-nav" aria-label="scikit-allel.yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example_3" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_3" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#dadi-cliyaml" class="md-nav__link">
    <span class="md-ellipsis">
      dadi-cli.yaml
    </span>
  </a>
  
    <nav class="md-nav" aria-label="dadi-cli.yaml">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#example_4" class="md-nav__link">
    <span class="md-ellipsis">
      Example
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#parameters_4" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#important-notes" class="md-nav__link">
    <span class="md-ellipsis">
      Important Notes
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="configuration">Configuration</h1>
<p>Selscape uses multiple YAML configuration files to control different aspects of the analysis:</p>
<table>
<thead>
<tr>
<th>File</th>
<th>Purpose</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>config/main.yaml</code></td>
<td>Core settings (species, populations, data paths)</td>
</tr>
<tr>
<td><code>config/selscan.yaml</code></td>
<td>selscan parameters</td>
</tr>
<tr>
<td><code>config/betascan.yaml</code></td>
<td>BetaScan parameters</td>
</tr>
<tr>
<td><code>config/scikit-allel.yaml</code></td>
<td>scikit-allel parameters</td>
</tr>
<tr>
<td><code>config/dadi-cli.yaml</code></td>
<td>dadi-cli parameters</td>
</tr>
<tr>
<td><code>config/slurm/config.yaml</code></td>
<td>HPC cluster settings (optional)</td>
</tr>
</tbody>
</table>
<h2 id="mainyaml">main.yaml</h2>
<h3 id="example">Example</h3>
<div class="codehilite"><pre><span></span><code><span class="c1"># Species identification</span>
<span class="nt">species</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;Human&quot;</span>
<span class="nt">tax_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">9606</span>
<span class="nt">ref_genome</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;hg38&quot;</span>
<span class="nt">ploidy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>

<span class="c1"># Populations to analyze</span>
<span class="nt">populations</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">YRI</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">CHS</span>

<span class="c1"># VCF file configuration</span>
<span class="nt">data_folder</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;examples/data/Human/raw&quot;</span>
<span class="nt">vcf_prefix</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;full_chr&quot;</span>
<span class="nt">vcf_suffix</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;.vcf.gz&quot;</span>
<span class="c1"># Files should be named: {vcf_prefix}{chromosome}{vcf_suffix}</span>
<span class="c1"># Example: full_chr21.vcf.gz</span>

<span class="nt">metadata</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;examples/data/Human/metadata/example_metadata.txt&quot;</span>

<span class="c1"># Chromosomes to analyze</span>
<span class="nt">chromosomes</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">20</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">21</span>

<span class="c1"># Ancestral alleles for polarization (optional but recommended)</span>
<span class="nt">anc_alleles</span><span class="p">:</span>
<span class="w">  </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;examples/data/Human/ancestral_alleles/homo_sapiens_ancestor_GRCh38&quot;</span>
<span class="w">  </span><span class="nt">prefix</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;homo_sapiens_ancestor_chr&quot;</span>
<span class="w">  </span><span class="c1"># Files: {prefix}{chromosome}.bed.gz</span>

<span class="c1"># Annotation files</span>
<span class="nt">genome_annotation</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;examples/data/Human/annotation/Human.gtf.gz&quot;</span>
<span class="nt">gene2go</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;examples/data/Human/annotation/gene2go.gz&quot;</span>

<span class="c1"># Quality control</span>
<span class="nt">hwe_pvalue</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.001</span><span class="w">  </span><span class="c1"># Hardy-Weinberg equilibrium threshold</span>

<span class="c1"># Repeat regions (optional)</span>
<span class="nt">rmsk</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;examples/data/Human/repeats/hg38.rmsk.autosomes.bed&quot;</span>
<span class="nt">seg_dup</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;examples/data/Human/repeats/hg38.seg.dups.autosomes.bed&quot;</span>
<span class="nt">sim_rep</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;examples/data/Human/repeats/hg38.simple.repeats.autosomes.bed&quot;</span>

<span class="c1"># Method configuration files</span>
<span class="nt">selscan_config</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;config/selscan.yaml&quot;</span>
<span class="nt">betascan_config</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;config/betascan.yaml&quot;</span>
<span class="nt">dadi_config</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;config/dadi-cli.yaml&quot;</span>
<span class="nt">scikit_allel_config</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;config/scikit-allel.yaml&quot;</span>
</code></pre></div>

<h3 id="parameters">Parameters</h3>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Type</th>
<th>Description</th>
<th>Required</th>
</tr>
</thead>
<tbody>
<tr>
<td>species</td>
<td>string</td>
<td>Species name</td>
<td>Y</td>
</tr>
<tr>
<td>tax_id</td>
<td>integer</td>
<td>NCBI taxonomy ID</td>
<td>Y</td>
</tr>
<tr>
<td>ref_genome</td>
<td>string</td>
<td>Reference genome build</td>
<td>Y</td>
</tr>
<tr>
<td>anc_alleles.path</td>
<td>string</td>
<td>Base folder for ancestral allele information</td>
<td>N</td>
</tr>
<tr>
<td>anc_alleles.prefix</td>
<td>string</td>
<td>Ancestral allele information filename prefix</td>
<td>N</td>
</tr>
<tr>
<td>genome_annotation</td>
<td>string</td>
<td>Path to genome annotation GTF file</td>
<td>Y</td>
</tr>
<tr>
<td>gene2go</td>
<td>string</td>
<td>Path to gene2go mapping file</td>
<td>Y</td>
</tr>
<tr>
<td>hwe_pvalue</td>
<td>float</td>
<td>Hardy-Weinberg equilibrium p-value threshold</td>
<td>Y</td>
</tr>
<tr>
<td>hwe_pvalue</td>
<td>float</td>
<td>Hardy-Weinberg equilibrium p-value threshold</td>
<td>Y</td>
</tr>
<tr>
<td>rmsk</td>
<td>string</td>
<td>Path to repeat masker BED file</td>
<td>N</td>
</tr>
<tr>
<td>seg_dup</td>
<td>string</td>
<td>Path to segmental duplications BED file</td>
<td>N</td>
</tr>
<tr>
<td>sim_rep</td>
<td>string</td>
<td>Path to simple repeats BED file</td>
<td>N</td>
</tr>
<tr>
<td>ploidy</td>
<td>integer</td>
<td>Organism ploidy</td>
<td>Y</td>
</tr>
<tr>
<td>populations</td>
<td>list</td>
<td>List of population IDs to analyze</td>
<td>Y</td>
</tr>
<tr>
<td>data_folder</td>
<td>string</td>
<td>Base folder for input VCF files</td>
<td>Y</td>
</tr>
<tr>
<td>vcf_prefix</td>
<td>string</td>
<td>VCF filename prefix</td>
<td>Y</td>
</tr>
<tr>
<td>vcf_suffix</td>
<td>string</td>
<td>VCF filename suffix</td>
<td>Y</td>
</tr>
<tr>
<td>metadata</td>
<td>string</td>
<td>Path to sample metadata file</td>
<td>Y</td>
</tr>
<tr>
<td>chromosomes</td>
<td>list</td>
<td>List of chromosomes to analyze</td>
<td>Y</td>
</tr>
<tr>
<td>betascan_config</td>
<td>string</td>
<td>Path to BetaScan configuration file</td>
<td>Y</td>
</tr>
<tr>
<td>selscan_config</td>
<td>string</td>
<td>Path to selscan configuration file</td>
<td>Y</td>
</tr>
<tr>
<td>dadi_config</td>
<td>string</td>
<td>Path to dadi-cli configuration file</td>
<td>Y</td>
</tr>
<tr>
<td>scikit_allel_config</td>
<td>string</td>
<td>Path to scikit-allel configuration file</td>
<td>Y</td>
</tr>
</tbody>
</table>
<h2 id="betascanyaml">betascan.yaml</h2>
<h3 id="example_1">Example</h3>
<div class="codehilite"><pre><span></span><code><span class="c1"># Data type</span>
<span class="nt">unfolded</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span><span class="w">  </span><span class="c1"># Use polarized data (requires ancestral alleles)</span>

<span class="c1"># Core allele frequency parameter</span>
<span class="nt">core_frq</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.15</span>

<span class="c1"># Allele frequency filters</span>
<span class="nt">min_af</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.05</span><span class="w">  </span><span class="c1"># Minimum allele frequency</span>
<span class="nt">max_af</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.95</span><span class="w">  </span><span class="c1"># Maximum allele frequency</span>

<span class="c1"># Statistical thresholds</span>
<span class="nt">top_proportion</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.0005</span><span class="w">  </span><span class="c1"># Top proportion for outlier identification</span>

<span class="c1"># Manhattan plot settings</span>
<span class="nt">manhattan_plot_width</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">640</span>
<span class="nt">manhattan_plot_height</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">240</span>
<span class="nt">manhattan_plot_color1</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;#56B4E9&quot;</span>
<span class="nt">manhattan_plot_color2</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;#F0E442&quot;</span>
</code></pre></div>

<h3 id="parameters_1">Parameters</h3>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Type</th>
<th>Description</th>
<th>Required</th>
</tr>
</thead>
<tbody>
<tr>
<td>unfolded</td>
<td>boolean</td>
<td>Use polarized/unfolded allele frequency spectrum</td>
<td>Y</td>
</tr>
<tr>
<td>core_frq</td>
<td>float</td>
<td>Core allele frequency parameter</td>
<td>Y</td>
</tr>
<tr>
<td>min_af</td>
<td>float</td>
<td>Minimum allele frequency</td>
<td>Y</td>
</tr>
<tr>
<td>max_af</td>
<td>float</td>
<td>Maximum allele frequency</td>
<td>Y</td>
</tr>
<tr>
<td>top_proportion</td>
<td>float</td>
<td>Top proportion for outlier identification</td>
<td>Y</td>
</tr>
</tbody>
</table>
<h2 id="selscanyaml">selscan.yaml</h2>
<h3 id="example_2">Example</h3>
<div class="codehilite"><pre><span></span><code><span class="c1"># Data type</span>
<span class="nt">unphased</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">true</span><span class="w">  </span><span class="c1"># Set to false if you have phased data</span>

<span class="c1"># Within-population statistics</span>
<span class="nt">wp_stats</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">ihs</span><span class="w">   </span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">nsl</span><span class="w">    </span>

<span class="c1"># Cross-population statistics  </span>
<span class="nt">xp_stats</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">xpehh</span><span class="w">  </span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">xpnsl</span>

<span class="c1"># Minor allele frequency threshold</span>
<span class="nt">maf</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.05</span>

<span class="c1"># Statistical thresholds</span>
<span class="nt">top_proportion</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.0005</span>

<span class="c1"># Manhattan plot settings</span>
<span class="nt">manhattan_plot_width</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">640</span>
<span class="nt">manhattan_plot_height</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">240</span>
<span class="nt">manhattan_plot_color1</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;#56B4E9&quot;</span>
<span class="nt">manhattan_plot_color2</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;#F0E442&quot;</span>
</code></pre></div>

<h3 id="parameters_2">Parameters</h3>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Type</th>
<th>Description</th>
<th>Required</th>
</tr>
</thead>
<tbody>
<tr>
<td>unphased</td>
<td>boolean</td>
<td>Use unphased (true) or phased (false) data</td>
<td>Y</td>
</tr>
<tr>
<td>wp_stats</td>
<td>list</td>
<td>Within-population statistics to compute</td>
<td>Y</td>
</tr>
<tr>
<td>xp_stats</td>
<td>list</td>
<td>Cross-population statistics to compute</td>
<td>Y</td>
</tr>
<tr>
<td>maf</td>
<td>float</td>
<td>Minor allele frequency threshold</td>
<td>Y</td>
</tr>
<tr>
<td>top_proportion</td>
<td>float</td>
<td>Top proportion for outlier identification</td>
<td>Y</td>
</tr>
</tbody>
</table>
<h2 id="scikit-allelyaml">scikit-allel.yaml</h2>
<h3 id="example_3">Example</h3>
<div class="codehilite"><pre><span></span><code><span class="c1"># Within-population statistics</span>
<span class="nt">wp_stats</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">windowed_tajima_d</span><span class="w">   </span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">moving_tajima_d</span><span class="w">  </span>

<span class="c1"># Windowed approach (base pair windows)</span>
<span class="nt">windowed_window_sizes</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">100_000</span><span class="p p-Indicator">]</span>
<span class="nt">windowed_step_size_ratios</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">1</span><span class="p p-Indicator">]</span><span class="w">  </span><span class="c1"># 1 = non-overlapping windows</span>

<span class="c1"># Moving approach (SNP windows)</span>
<span class="nt">moving_window_sizes</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">100</span><span class="p p-Indicator">]</span>
<span class="nt">moving_step_size_ratios</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">1</span><span class="p p-Indicator">]</span>

<span class="c1"># Cross-population statistics </span>
<span class="nt">xp_stats</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">delta_moving_tajima_d</span>

<span class="c1"># XP moving approach (SNP windows)</span>
<span class="nt">delta_moving_window_sizes</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">100</span><span class="p p-Indicator">]</span>
<span class="nt">delta_moving_step_size_ratios</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="nv">1</span><span class="p p-Indicator">]</span>

<span class="c1"># Statistical thresholds</span>
<span class="nt">top_proportion</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0.05</span>

<span class="c1"># Manhattan plot settings</span>
<span class="nt">manhattan_plot_width</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">640</span>
<span class="nt">manhattan_plot_height</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">240</span>
<span class="nt">manhattan_plot_color1</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;#56B4E9&quot;</span>
<span class="nt">manhattan_plot_color2</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;#F0E442&quot;</span>
</code></pre></div>

<h3 id="parameters_3">Parameters</h3>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Type</th>
<th>Description</th>
<th>Required</th>
</tr>
</thead>
<tbody>
<tr>
<td>wp_stats</td>
<td>list</td>
<td>Within-population statistics to compute</td>
<td>Y</td>
</tr>
<tr>
<td>xp_stats</td>
<td>list</td>
<td>Cross-population statistics to compute</td>
<td>Y</td>
</tr>
<tr>
<td>wtjd_window_sizes</td>
<td>list</td>
<td>Window sizes in base pairs for windowed tajima's d (wtjd) statistic</td>
<td>Y</td>
</tr>
<tr>
<td>wtjd_step_size_ratios</td>
<td>list</td>
<td>Step size as fraction of window in base pair for wtjd statistic</td>
<td>Y</td>
</tr>
<tr>
<td>mtjd_window_sizes</td>
<td>list</td>
<td>Window sizes in SNPs for moving tajima's d (mtjd) statistic</td>
<td>Y</td>
</tr>
<tr>
<td>mtjd_step_size_ratios</td>
<td>list</td>
<td>Step size as fraction of window in SNPs for mtjd statistic</td>
<td>Y</td>
</tr>
<tr>
<td>dtjd_moving_window_sizes</td>
<td>list</td>
<td>Window sizes in SNPs for xp moving delta tajima's d (dtjd) statistic</td>
<td>Y</td>
</tr>
<tr>
<td>dtjd_moving_step_size_ratios</td>
<td>list</td>
<td>Step size as fraction of window in SNPs for cross population dtjd statistic</td>
<td>Y</td>
</tr>
<tr>
<td>top_proportion</td>
<td>float</td>
<td>Top proportion for outlier identification</td>
<td>Y</td>
</tr>
</tbody>
</table>
<h2 id="dadi-cliyaml">dadi-cli.yaml</h2>
<h3 id="example_4">Example</h3>
<div class="codehilite"><pre><span></span><code><span class="c1"># Data type</span>
<span class="nt">unfolded</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">True</span>

<span class="c1"># Demographic models</span>
<span class="nt">demog_1d</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">two_epoch</span>
<span class="nt">demog_1d_p0</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;0.5</span><span class="nv"> </span><span class="s">5</span><span class="nv"> </span><span class="s">0.5&quot;</span>
<span class="nt">demog_1d_ub</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;10</span><span class="nv"> </span><span class="s">10</span><span class="nv"> </span><span class="s">1&quot;</span>
<span class="nt">demog_1d_lb</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;10e-5</span><span class="nv"> </span><span class="s">10e-5</span><span class="nv"> </span><span class="s">0&quot;</span>

<span class="c1"># DFE models</span>
<span class="nt">dfe_1d</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">lognormal</span>
<span class="nt">dfe_1d_p0</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;5</span><span class="nv"> </span><span class="s">5</span><span class="nv"> </span><span class="s">0.5&quot;</span>
<span class="nt">dfe_1d_ub</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;100</span><span class="nv"> </span><span class="s">1000</span><span class="nv"> </span><span class="s">1&quot;</span>
<span class="nt">dfe_1d_lb</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;10e-5</span><span class="nv"> </span><span class="s">10e-5</span><span class="nv"> </span><span class="s">0&quot;</span>

<span class="c1"># Grid sizes</span>
<span class="nt">dm_grid_size</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;300</span><span class="nv"> </span><span class="s">400</span><span class="nv"> </span><span class="s">500&quot;</span>
<span class="nt">dfe_grid_size</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;300</span><span class="nv"> </span><span class="s">400</span><span class="nv"> </span><span class="s">500&quot;</span>

<span class="c1"># Gamma points for generating caches</span>
<span class="nt">gamma_pts</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span>

<span class="c1"># Ratio of non-synonymous/synonymous mutation rates</span>
<span class="nt">ratio</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2.31</span>

<span class="c1"># Number of optimization runs</span>
<span class="nt">optimizations</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">50</span>

<span class="c1"># Bootstrap parameters</span>
<span class="nt">bootstrap_replicates</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">100</span>
<span class="nt">chunk_size</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1000000</span>

<span class="c1"># Frequency spectrum flags</span>
<span class="nt">mask_singletons</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
</code></pre></div>

<h3 id="parameters_4">Parameters</h3>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Type</th>
<th>Description</th>
<th>Required</th>
</tr>
</thead>
<tbody>
<tr>
<td>unfolded</td>
<td>boolean</td>
<td>Use polarized/unfolded allele frequency spectrum</td>
<td>Y</td>
</tr>
<tr>
<td>demog_1d</td>
<td>string</td>
<td>1D demographic model name</td>
<td>Y</td>
</tr>
<tr>
<td>demog_1d_p0</td>
<td>string</td>
<td>Initial parameters for 1D demographic model</td>
<td>Y</td>
</tr>
<tr>
<td>demog_1d_ub</td>
<td>string</td>
<td>Upper bounds for 1D demographic model parameters</td>
<td>Y</td>
</tr>
<tr>
<td>demog_1d_lb</td>
<td>string</td>
<td>Lower bounds for 1D demographic model parameters</td>
<td>Y</td>
</tr>
<tr>
<td>dfe_1d</td>
<td>string</td>
<td>1D DFE model name</td>
<td>Y</td>
</tr>
<tr>
<td>dfe_1d_p0</td>
<td>string</td>
<td>Initial parameters for 1D DFE model</td>
<td>Y</td>
</tr>
<tr>
<td>dfe_1d_ub</td>
<td>string</td>
<td>Upper bounds for 1D DFE model parameters</td>
<td>Y</td>
</tr>
<tr>
<td>dfe_1d_lb</td>
<td>string</td>
<td>Lower bounds for 1D DFE model parameters</td>
<td>Y</td>
</tr>
<tr>
<td>dm_grid_size</td>
<td>string</td>
<td>Grid sizes for demographic inference</td>
<td>Y</td>
</tr>
<tr>
<td>dfe_grid_size</td>
<td>string</td>
<td>Grid for DFE inference</td>
<td>Y</td>
</tr>
<tr>
<td>gamma_pts</td>
<td>integer</td>
<td>Number of gamma points for cache generation</td>
<td>Y</td>
</tr>
<tr>
<td>ratio</td>
<td>float</td>
<td>Ratio of non-synonymous to synonymous mutation rates</td>
<td>Y</td>
</tr>
<tr>
<td>optimizations</td>
<td>integer</td>
<td>Number of optimization runs</td>
<td>Y</td>
</tr>
<tr>
<td>bootstrap_replicates</td>
<td>integer</td>
<td>Number of bootstrap replicates</td>
<td>Y</td>
</tr>
<tr>
<td>chunk_size</td>
<td>integer</td>
<td>Chunk size for bootstrapping</td>
<td>Y</td>
</tr>
<tr>
<td>mask_singletons</td>
<td>boolean</td>
<td>Whether to mask singleton variants</td>
<td>Y</td>
</tr>
</tbody>
</table>
<h2 id="important-notes">Important Notes</h2>
<ol>
<li>
<p><strong>Ancestral Alleles</strong>: If <code>anc_alleles</code> is not provided, polarization-dependent analyses (selscan, betascan with unfolded=True, dadi with unfolded=True) will not run.</p>
</li>
<li>
<p><strong>VCF File Naming</strong>: VCF files must follow the pattern: <code>{data_folder}/{vcf_prefix}{chromosome}{vcf_suffix}</code></p>
</li>
<li>Example with <code>vcf_prefix: "full_chr"</code>: <code>examples/data/Human/raw/full_chr21.vcf.gz</code></li>
<li>Example with <code>vcf_prefix: "chr"</code>: <code>resources/data/chr21.vcf.gz</code></li>
<li>
<p>Example with <code>vcf_prefix: ""</code>: <code>resources/data/21.vcf.gz</code></p>
</li>
<li>
<p><strong>Metadata Format</strong>: Tab-separated file with columns: <code>Sample</code> and <code>Population</code></p>
</li>
<li>
<p><strong>Repeat Masking</strong>: Optional but recommended. Set to empty string or null to disable.</p>
</li>
<li>
<p><strong>Step Size Ratios</strong>: Values from 0 to 1 represent the fraction of window size to step. </p>
</li>
<li>Example: window size = 100, step size ratio = 0.1 â†’ step size = 10</li>
</ol>
<p>Please check <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html#configuration">Snakemake Configuration</a> documentation for additional information.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.525ec568.min.js"></script>
      
    
  </body>
</html>