
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://xin-huang.github.io/selscape/userguide/usage/">
      
      
        <link rel="prev" href="../..">
      
      
        <link rel="next" href="../configuration/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.5.39">
    
    
      
        <title>Usage - selscape</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.8c3ca2c6.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="teal" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#usage" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="selscape" class="md-header__button md-logo" aria-label="selscape" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            selscape
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Usage
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/xin-huang/selscape" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="selscape" class="md-nav__button md-logo" aria-label="selscape" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    selscape
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/xin-huang/selscape" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.6.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    User Guide
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            User Guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Usage
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Usage
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#quick-start-with-example-data" class="md-nav__link">
    <span class="md-ellipsis">
      Quick Start with Example Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Quick Start with Example Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-download-example-data" class="md-nav__link">
    <span class="md-ellipsis">
      1. Download Example Data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-run-analysis-on-example-data" class="md-nav__link">
    <span class="md-ellipsis">
      2. Run Analysis on Example Data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analyzing-your-own-data" class="md-nav__link">
    <span class="md-ellipsis">
      Analyzing Your Own Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Analyzing Your Own Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-prepare-your-data" class="md-nav__link">
    <span class="md-ellipsis">
      1. Prepare Your Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Prepare Your Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#vcf-files" class="md-nav__link">
    <span class="md-ellipsis">
      VCF Files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#metadata-file" class="md-nav__link">
    <span class="md-ellipsis">
      Metadata File
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ancestral-alleles-optional" class="md-nav__link">
    <span class="md-ellipsis">
      Ancestral Alleles (Optional)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-configure-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      2. Configure Analysis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-run-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      3. Run Analysis
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-specific-analyses" class="md-nav__link">
    <span class="md-ellipsis">
      Running Specific Analyses
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generate-html-report" class="md-nav__link">
    <span class="md-ellipsis">
      Generate HTML Report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#clean-restart" class="md-nav__link">
    <span class="md-ellipsis">
      Clean Restart
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow-options" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow Options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#troubleshooting" class="md-nav__link">
    <span class="md-ellipsis">
      Troubleshooting
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../configuration/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../output/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Output
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#quick-start-with-example-data" class="md-nav__link">
    <span class="md-ellipsis">
      Quick Start with Example Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Quick Start with Example Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-download-example-data" class="md-nav__link">
    <span class="md-ellipsis">
      1. Download Example Data
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-run-analysis-on-example-data" class="md-nav__link">
    <span class="md-ellipsis">
      2. Run Analysis on Example Data
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#analyzing-your-own-data" class="md-nav__link">
    <span class="md-ellipsis">
      Analyzing Your Own Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Analyzing Your Own Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#1-prepare-your-data" class="md-nav__link">
    <span class="md-ellipsis">
      1. Prepare Your Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="1. Prepare Your Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#vcf-files" class="md-nav__link">
    <span class="md-ellipsis">
      VCF Files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#metadata-file" class="md-nav__link">
    <span class="md-ellipsis">
      Metadata File
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#ancestral-alleles-optional" class="md-nav__link">
    <span class="md-ellipsis">
      Ancestral Alleles (Optional)
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#2-configure-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      2. Configure Analysis
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#3-run-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      3. Run Analysis
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#running-specific-analyses" class="md-nav__link">
    <span class="md-ellipsis">
      Running Specific Analyses
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#generate-html-report" class="md-nav__link">
    <span class="md-ellipsis">
      Generate HTML Report
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#clean-restart" class="md-nav__link">
    <span class="md-ellipsis">
      Clean Restart
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#workflow-options" class="md-nav__link">
    <span class="md-ellipsis">
      Workflow Options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#troubleshooting" class="md-nav__link">
    <span class="md-ellipsis">
      Troubleshooting
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


<h1 id="usage">Usage</h1>
<p><code>selscape</code> provides example data to test the workflow, and supports analyzing your own datasets with flexible configuration.</p>
<h2 id="quick-start-with-example-data">Quick Start with Example Data</h2>
<h3 id="1-download-example-data">1. Download Example Data</h3>
<p><code>selscape</code> includes a workflow to download example data from the 1000 Genomes Project:</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># Download example data (YRI and CHS populations, chromosomes 20-21)</span>
snakemake<span class="w"> </span>-s<span class="w"> </span>examples/get_example_data.smk<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span>
</code></pre></div>

<p>This creates the following structure:</p>
<div class="codehilite"><pre><span></span><code><span class="nx">examples</span><span class="o">/</span><span class="nx">data</span><span class="o">/</span><span class="nx">Human</span><span class="o">/</span>
<span class="err">├──</span><span class="w"> </span><span class="nx">raw</span><span class="o">/</span>
<span class="err">│</span><span class="w">   </span><span class="err">├──</span><span class="w"> </span><span class="nx">full_chr20</span><span class="p">.</span><span class="nx">vcf</span><span class="p">.</span><span class="nx">gz</span>
<span class="err">│</span><span class="w">   </span><span class="err">├──</span><span class="w"> </span><span class="nx">full_chr20</span><span class="p">.</span><span class="nx">vcf</span><span class="p">.</span><span class="nx">gz</span><span class="p">.</span><span class="nx">tbi</span>
<span class="err">│</span><span class="w">   </span><span class="err">├──</span><span class="w"> </span><span class="nx">full_chr21</span><span class="p">.</span><span class="nx">vcf</span><span class="p">.</span><span class="nx">gz</span>
<span class="err">│</span><span class="w">   </span><span class="err">└──</span><span class="w"> </span><span class="nx">full_chr21</span><span class="p">.</span><span class="nx">vcf</span><span class="p">.</span><span class="nx">gz</span><span class="p">.</span><span class="nx">tbi</span>
<span class="err">├──</span><span class="w"> </span><span class="nx">ancestral_alleles</span><span class="o">/</span>
<span class="err">│</span><span class="w">   </span><span class="err">└──</span><span class="w"> </span><span class="nx">homo_sapiens_ancestor_GRCh38</span><span class="o">/</span>
<span class="err">├──</span><span class="w"> </span><span class="nx">repeats</span><span class="o">/</span>
<span class="err">│</span><span class="w">   </span><span class="err">├──</span><span class="w"> </span><span class="nx">hg38</span><span class="p">.</span><span class="nx">rmsk</span><span class="p">.</span><span class="nx">autosomes</span><span class="p">.</span><span class="nx">bed</span>
<span class="err">│</span><span class="w">   </span><span class="err">├──</span><span class="w"> </span><span class="nx">hg38</span><span class="p">.</span><span class="nx">seg</span><span class="p">.</span><span class="nx">dups</span><span class="p">.</span><span class="nx">autosomes</span><span class="p">.</span><span class="nx">bed</span>
<span class="err">│</span><span class="w">   </span><span class="err">└──</span><span class="w"> </span><span class="nx">hg38</span><span class="p">.</span><span class="nx">simple</span><span class="p">.</span><span class="nx">repeats</span><span class="p">.</span><span class="nx">autosomes</span><span class="p">.</span><span class="nx">bed</span>
<span class="err">├──</span><span class="w"> </span><span class="nx">annotation</span><span class="o">/</span>
<span class="err">│</span><span class="w">   </span><span class="err">├──</span><span class="w"> </span><span class="nx">Human</span><span class="p">.</span><span class="nx">gtf</span><span class="p">.</span><span class="nx">gz</span>
<span class="err">│</span><span class="w">   </span><span class="err">└──</span><span class="w"> </span><span class="nx">gene2go</span><span class="p">.</span><span class="nx">gz</span>
<span class="err">└──</span><span class="w"> </span><span class="nx">metadata</span><span class="o">/</span>
<span class="w">    </span><span class="err">└──</span><span class="w"> </span><span class="nx">example_metadata</span><span class="p">.</span><span class="nx">txt</span>
</code></pre></div>

<h3 id="2-run-analysis-on-example-data">2. Run Analysis on Example Data</h3>
<div class="codehilite"><pre><span></span><code><span class="c1"># Local execution (single core)</span>
snakemake<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span><span class="w"> </span>--configfile<span class="w"> </span>config/main.yaml

<span class="c1"># Local execution (multiple cores)</span>
snakemake<span class="w"> </span>-c<span class="w"> </span><span class="m">4</span><span class="w"> </span>--configfile<span class="w"> </span>config/main.yaml

<span class="c1"># HPC cluster (SLURM)</span>
snakemake<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span><span class="w"> </span>--configfile<span class="w"> </span>config/main.yaml<span class="w"> </span>--profile<span class="w"> </span>config/slurm/<span class="w"> </span>-j<span class="w"> </span><span class="m">50</span>
</code></pre></div>

<h2 id="analyzing-your-own-data">Analyzing Your Own Data</h2>
<h3 id="1-prepare-your-data">1. Prepare Your Data</h3>
<h4 id="vcf-files">VCF Files</h4>
<p>Organize your VCF files according to your chosen naming scheme. The configuration will tell Selscape how to find them.</p>
<p>Example structures:</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># Option 1: Chromosome in filename</span>
resources/data/
├──<span class="w"> </span>chr1.vcf.gz
├──<span class="w"> </span>chr1.vcf.gz.tbi
├──<span class="w"> </span>chr2.vcf.gz
└──<span class="w"> </span>chr2.vcf.gz.tbi

<span class="c1"># Option 2: Prefix + chromosome</span>
resources/data/
├──<span class="w"> </span>Sample_chr1.vcf.gz
├──<span class="w"> </span>Sample_chr1.vcf.gz.tbi
├──<span class="w"> </span>Sample_chr2.vcf.gz
└──<span class="w"> </span>Sample_chr2.vcf.gz.tbi
</code></pre></div>

<p><strong>Requirements:</strong>
- All VCF files must be bgzipped (<code>.vcf.gz</code>)
- All VCF files must be indexed (<code>.vcf.gz.tbi</code>)
- Biallelic SNPs only (can be filtered by selscape)</p>
<h4 id="metadata-file">Metadata File</h4>
<p>Create a tab-separated file mapping samples to populations:</p>
<div class="codehilite"><pre><span></span><code>Sample  Population
NA19001 YRI
NA19002 YRI
NA19003 YRI
HG00419 CHS
HG00420 CHS
HG00421 CHS
</code></pre></div>

<p><strong>Requirements:</strong>
- Header line: <code>Sample</code> and <code>Population</code>
- Tab-separated
- Sample IDs must match VCF file samples</p>
<h4 id="ancestral-alleles-optional">Ancestral Alleles (Optional)</h4>
<p>If available, ancestral alleles enable polarized analyses (selscan, unfolded BetaScan/dadi):</p>
<div class="codehilite"><pre><span></span><code>ancestral_alleles/
├──<span class="w"> </span>anc_chr1.bed.gz
├──<span class="w"> </span>anc_chr1.bed.gz.tbi
├──<span class="w"> </span>anc_chr2.bed.gz
└──<span class="w"> </span>anc_chr2.bed.gz.tbi
</code></pre></div>

<p><strong>Format:</strong> BED file with ancestral base in 4th column:</p>
<div class="codehilite"><pre><span></span><code>chr1    0   1   A
chr1    1   2   C
chr1    2   3   G
</code></pre></div>

<h3 id="2-configure-analysis">2. Configure Analysis</h3>
<p>Edit <code>config/main.yaml</code>:</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># Species identification</span>
<span class="nt">species</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;YourSpecies&quot;</span>
<span class="nt">tax_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">9606</span><span class="w">  </span><span class="c1"># Update for your species</span>
<span class="nt">ref_genome</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;your_ref&quot;</span>
<span class="nt">ploidy</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>

<span class="c1"># Populations to analyze</span>
<span class="nt">populations</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Pop1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">Pop2</span>

<span class="c1"># VCF file configuration - ADJUST TO YOUR NAMING SCHEME</span>
<span class="nt">data_folder</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;resources/data&quot;</span>
<span class="nt">vcf_prefix</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;chr&quot;</span><span class="w">  </span><span class="c1"># or &quot;Sample_chr&quot;, &quot;full_chr&quot;, etc.</span>
<span class="nt">vcf_suffix</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;.vcf.gz&quot;</span>

<span class="c1"># Sample metadata</span>
<span class="nt">metadata</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;resources/metadata/samples.txt&quot;</span>

<span class="c1"># Chromosomes to analyze</span>
<span class="nt">chromosomes</span><span class="p">:</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">1</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">3</span>
<span class="w">  </span><span class="c1"># ... add all chromosomes</span>

<span class="c1"># Ancestral alleles (optional)</span>
<span class="nt">anc_alleles</span><span class="p">:</span>
<span class="w">  </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;resources/ancestral_alleles&quot;</span>
<span class="w">  </span><span class="nt">prefix</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;anc_chr&quot;</span>

<span class="c1"># Annotation files</span>
<span class="nt">genome_annotation</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;resources/annotation/genome.gtf.gz&quot;</span>
<span class="nt">gene2go</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;resources/annotation/gene2go.gz&quot;</span>

<span class="c1"># Repeat regions (optional but recommended)</span>
<span class="nt">rmsk</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;resources/repeats/repeats.bed&quot;</span>
<span class="nt">seg_dup</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;resources/repeats/seg_dups.bed&quot;</span>
<span class="nt">sim_rep</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;resources/repeats/simple_repeats.bed&quot;</span>

<span class="c1"># Analysis tool configuration</span>
<span class="nt">selscan_config</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;config/selscan.yaml&quot;</span>
<span class="nt">betascan_config</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;config/betascan.yaml&quot;</span>
<span class="nt">dadi_config</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;config/dadi-cli.yaml&quot;</span>
<span class="nt">scikit_allel_config</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;config/scikit-allel.yaml&quot;</span>
</code></pre></div>

<h3 id="3-run-analysis">3. Run Analysis</h3>
<div class="codehilite"><pre><span></span><code><span class="c1"># Preview what will be executed (dry run)</span>
snakemake<span class="w"> </span>-np<span class="w"> </span>--configfile<span class="w"> </span>/path/to/your/main.yaml

<span class="c1"># Local execution</span>
snakemake<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span><span class="w"> </span>--configfile<span class="w"> </span>/path/to/your/main.yaml

<span class="c1"># HPC cluster (SLURM)</span>
snakemake<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span><span class="w"> </span>--configfile<span class="w"> </span>/path/to/your/main.yaml<span class="w"> </span>--profile<span class="w"> </span>config/slurm/<span class="w"> </span>--j<span class="w"> </span><span class="m">50</span>
</code></pre></div>

<h2 id="running-specific-analyses">Running Specific Analyses</h2>
<p>You can run only specific methods by requesting specific output files:</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># Only BetaScan</span>
snakemake<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>results/balancing_selection/betascan/Human/YRI/m_0.15/YRI.hg38.m_0.15.b1.top_0.0005.scores.png

<span class="c1"># Only selscan (requires ancestral alleles)</span>
snakemake<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>results/positive_selection/selscan/Human/1pop/YRI/ihs_0.05/YRI.normalized.ihs.maf_0.05.top_0.0005.scores.png

<span class="c1"># Only scikit-allel</span>
snakemake<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>results/positive_selection/scikit-allel/Human/1pop/YRI/windowed_tajima_d/100000_1/YRI.windowed_tajima_d.top_0.05.scores.png

<span class="c1"># Only dadi-cli</span>
snakemake<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>results/dadi/Human/dfe/YRI/plots/YRI.hg38.two_epoch.lognormal.fitted.png
</code></pre></div>

<h2 id="generate-html-report">Generate HTML Report</h2>
<p>Snakemake can automatically generate an interactive HTML report with main results, plots, and tables:</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># Generate report after workflow completion</span>
snakemake<span class="w"> </span>--report<span class="w"> </span>--configfile<span class="w"> </span>/path/to/your/main.yaml

<span class="c1"># Customize report location and file name</span>
snakemake<span class="w"> </span>--report<span class="w"> </span>/path/to/my_report.html<span class="w"> </span>--configfile<span class="w"> </span>/path/to/your/main.yaml
</code></pre></div>

<h2 id="clean-restart">Clean Restart</h2>
<p>To remove all results and start fresh:</p>
<div class="codehilite"><pre><span></span><code><span class="c1"># Remove all results (keeps raw data)</span>
rm<span class="w"> </span>-rf<span class="w"> </span>results/<span class="w"> </span>logs/

<span class="c1"># Remove everything including downloaded data</span>
rm<span class="w"> </span>-rf<span class="w"> </span>resources/<span class="w"> </span>results/<span class="w"> </span>logs/
</code></pre></div>

<h2 id="workflow-options">Workflow Options</h2>
<table>
<thead>
<tr>
<th>Command</th>
<th>Purpose</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>-c 1</code> or <code>--cores 1</code></td>
<td>Use 1 core (safest for testing)</td>
</tr>
<tr>
<td><code>-c 4</code></td>
<td>Use 4 parallel jobs (local)</td>
</tr>
<tr>
<td><code>-j or --jobs 50</code></td>
<td>Submit up to 50 jobs (cluster)</td>
</tr>
<tr>
<td><code>--profile config/slurm/</code></td>
<td>Use SLURM cluster configuration</td>
</tr>
<tr>
<td><code>-n</code> or <code>--dry-run</code></td>
<td>Preview what will be executed</td>
</tr>
<tr>
<td><code>-p</code></td>
<td>Print shell commands</td>
</tr>
<tr>
<td><code>--forceall</code></td>
<td>Re-run all steps</td>
</tr>
<tr>
<td><code>--unlock</code></td>
<td>Unlock working directory after crash</td>
</tr>
</tbody>
</table>
<h2 id="troubleshooting">Troubleshooting</h2>
<p><strong>Workflow locked:</strong></p>
<div class="codehilite"><pre><span></span><code>snakemake<span class="w"> </span>--unlock
</code></pre></div>

<p><strong>Re-run specific rule:</strong></p>
<div class="codehilite"><pre><span></span><code>snakemake<span class="w"> </span>--forcerun<span class="w"> </span>&lt;rule_name&gt;<span class="w"> </span>-c<span class="w"> </span><span class="m">1</span>
</code></pre></div>

<p><strong>Check what will run:</strong></p>
<div class="codehilite"><pre><span></span><code>snakemake<span class="w"> </span>-n<span class="w"> </span>-p
</code></pre></div>

<p>For more information, visit the <a href="https://snakemake.readthedocs.io/en/stable/index.html">Snakemake documentation</a>.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": [], "search": "../../assets/javascripts/workers/search.6ce7567c.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.525ec568.min.js"></script>
      
    
  </body>
</html>